import{_ as a,c as t,b as s,o}from"./app-JTHw23u4.js";const r={};function n(i,e){return o(),t("div",null,e[0]||(e[0]=[s(`<h1 id="storing-data-on-the-cluster" tabindex="-1"><a class="header-anchor" href="#storing-data-on-the-cluster"><span>Storing Data on the Cluster</span></a></h1><p>The cluster has an attached storage system onto which you can save your data. There are three main areas into which you can put data and which are appropriate for different kinds of data. All of the areas are situated under the <code>/bi</code> directory and you shouldn’t try to save data outside of these locations.</p><p>For all of the storage areas please try to manage the data you create. We have a relatively large storage area, but with the size of datasets we are now creating it will start to fill up surprisingly quickly. At the moment we do not put limits on the amount of space which can be taken up by individual users or groups, but this relies on people managing their storage usage and cleaning up data which is no longer required.</p><h2 id="home-shares" tabindex="-1"><a class="header-anchor" href="#home-shares"><span>Home shares</span></a></h2><p>Every user has a home directory under <code>/bi/home/[username]</code> and this is where you’ll start each session on the cluster. The home directory is suitable for relatively small files which are only relevant to you. Examples might be scripts that you’re working on or configuration or results files.</p><h2 id="group-shares" tabindex="-1"><a class="header-anchor" href="#group-shares"><span>Group shares</span></a></h2><p>Most of your important data should go into your group share which is situated under <code>/bi/group</code>. If you’re not sure which group you’re in you can just run <code>groups</code> on the command line which will tell you.</p><div class="language-text line-numbers-mode" data-highlighter="prismjs" data-ext="text" data-title="text"><pre><code><span class="line">[andrewss@capstone ~]$ groups andrewss</span>
<span class="line">andrewss : bioinf</span>
<span class="line"></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div></div></div><p>In this case you can see that I am part of the <code>bioinf</code> group, so my group share is at <code>/bi/group/bioinf</code>.</p><p>How files inside each group share are organised is left up to the head of that group so you’ll need to discuss with the other members of your group how you are going to manage this.</p><p>All data saved in either your home directory or the group share will be backed up to a second storage cluster at regular intervals. The backups are filesystem snapshots so although they provide some resilience they do not provide long term backups. If you accidentally delete something from one of these areas you need to get it restored as quickly as possible as the snapshots will roll over and the data will be permanently lost.</p><h2 id="scratch-area" tabindex="-1"><a class="header-anchor" href="#scratch-area"><span>Scratch area</span></a></h2><p>For data which doesn’t need to be backed up there is a separate storage area which you can use and you should try to use this area where possible so that we don’t end up wasting space in the backups for data which can easily be recreated. This scratch area can be found under <code>/bi/scratch</code> and its permissions are open so that anyone can write data into this area.</p><h2 id="other-data-areas" tabindex="-1"><a class="header-anchor" href="#other-data-areas"><span>Other data areas</span></a></h2><p>There are a couple of other storage areas on the cluster which can be useful for retrieving data rather than having to store a copy in your own space.</p><h3 id="sequencing-data" tabindex="-1"><a class="header-anchor" href="#sequencing-data"><span>Sequencing data</span></a></h3><p>If you have run samples through the Babraham sequencing facility, or had them run elsewhere and imported to Babraham, then you can access the data for these samples in the <code>/bi/sequencing</code> share. The <code>/bi/sequencing</code> share is a virtual file system, so every user will see something different in there. You should see a list of folders relating to the samples you submitted, or samples from other people to which you’ve been given permission in sierra. Under these folders you’ll see the lanes of sequencing run for each sample, and under these the unaligned and aligned data from those runs.</p><p>The sequencing share is read-only so you can’t modify the data in there. We recommend therefore that you either read the data you need directly out of this share or, if you want to refer to those files within your own area, that you create a symlink from <code>/bi/sequencing</code> to your own area. To create a symlink you use the <code>ln -s</code> command which will make a virtual file which you can use just like a normal file, but doesn’t actually take up any space in the file system.</p><p>If you find that you don&#39;t see your data in <code>/bi/sequencing</code> but just a text file called <code>README_UNREGISTERED.txt</code> then contact the bioinformatics group who can link your sierra account to your cluster account to make this feature work.</p><h3 id="imaging-data" tabindex="-1"><a class="header-anchor" href="#imaging-data"><span>Imaging Data</span></a></h3><p>If you have data stored on the ImageDataStore then this will be visible on the cluster under <code>/bi/imaging</code>. This is another virtual filesystem which only shows your data to you. Access to this data is read-only.</p><p>If you can’t see your imaging data then contact bioinformatics and we can configure the appropriate linkage between your user account and the folder(s) that you own on the imaging system.</p><h3 id="cached-public-datasets" tabindex="-1"><a class="header-anchor" href="#cached-public-datasets"><span>Cached public datasets</span></a></h3><p>To try to cut down on the number of copies of popular public datasets we stored on the cluster we have created a common area for these downloads which is /bi/pubcache. Although you can write into this area, we would prefer that any new datasets are requested by using the Labrador web system (at <a href="https://bioinfdev/labrador" target="_blank" rel="noopener noreferrer">https://bioinfdev/labrador</a>) so that everyone can see where they are. As for the sequencing share you can use symlinks to make the data under <code>/bi/pubcache</code> appear within your own working area.</p>`,24)]))}const d=a(r,[["render",n],["__file","data.html.vue"]]),c=JSON.parse('{"path":"/docs/data.html","title":"Storing Data on the Cluster","lang":"en-US","frontmatter":{},"headers":[{"level":2,"title":"Home shares","slug":"home-shares","link":"#home-shares","children":[]},{"level":2,"title":"Group shares","slug":"group-shares","link":"#group-shares","children":[]},{"level":2,"title":"Scratch area","slug":"scratch-area","link":"#scratch-area","children":[]},{"level":2,"title":"Other data areas","slug":"other-data-areas","link":"#other-data-areas","children":[{"level":3,"title":"Sequencing data","slug":"sequencing-data","link":"#sequencing-data","children":[]},{"level":3,"title":"Imaging Data","slug":"imaging-data","link":"#imaging-data","children":[]},{"level":3,"title":"Cached public datasets","slug":"cached-public-datasets","link":"#cached-public-datasets","children":[]}]}],"git":{"updatedTime":1727443175000,"contributors":[{"name":"Simon Andrews","email":"simon@proteo.me.uk","commits":2}]},"filePathRelative":"docs/data.md"}');export{d as comp,c as data};
